.github/workflows/release.yml

```diff
@@ -0,0 +1,107 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+name: Release
+
+on:
+  push:
+    branches:
+      - "maint-[0-9]+.[0-9]+.[0-9]+"
+      - "release-[0-9]+.[0-9]+.[0-9]+"
+      - "release-[0-9]+.[0-9]+.[0-9]+-rc[0-9]+"
+    tags:
+      # Trigger workflow when a tag whose name matches either
+      # pattern "apache-arrow-{MAJOR}.{MINOR}.{PATCH}" or pattern
+      # "apache-arrow-{MAJOR}.{MINOR}.{PATCH}-rc{RC_NUM}" is pushed.
+      - "apache-arrow-[0-9]+.[0-9]+.[0-9]+"
+      - "apache-arrow-[0-9]+.[0-9]+.[0-9]+-rc[0-9]+"
+
+permissions:
+  contents: write
+
+env:
+  GH_TOKEN: ${{ github.token }}
+
+jobs:
+  publish:
```

Comment:
How about creating two jobs for release and release candidate instead of putting `if` to each step?

```yaml
jobs:
  publish-candidate:
    if: |
      contains(${GITHUB_REF_NAME}, 'rc')
    ...
  publish:
    if: |
      !contains(${GITHUB_REF_NAME}, 'rc')
    ...
```

Reply:
Having two jobs might be better after all, I thought we could join it in one job but the required steps are just slightly different enough that some degree of duplication makes it much easier to read (I can not read through this without forgetting which step applies when xD :see_no_evil: ).

Reply:
Sounds good to me. There's some similarity between the two jobs, but I suspect they'll start to diverge more as more steps are added.